### PR TITLE
Skip truncated links

### LINKED ISSUE #39: Avoid linkifying truncated URLs
Truncated URLs (those ending with `…`) are usually broken and should not be linkified.

### PR DESCRIPTION
**Pull request by Katsute**
Closes #39 

~`getAsString` skip linkifying if the input ends with `…` and the URL is next to that `…`.~

~`getAsDocumentFragment` skip linkifying if the URL is the second to last item and the last item is `…`.~

Skips linkifying any URL followed by `…` or `...`

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -19,12 +19,13 @@ const linkify = (href, options) => createHtmlElement({
 // Get DOM node from HTML
 const domify = html => document.createRange().createContextualFragment(html);
 
-const getAsString = (string, options) => string.replace(urlRegex(), match => linkify(match, options));
+const getAsString = (string, options) => string.replace(urlRegex(), (match, _, offset) => offset + match.length + 1 === string.length && string.charAt(string.length - 1) === '…' ? match : linkify(match, options));
```
Comment (sindresorhus) [inline index.js]: Use `String#endsWith('…')`

File: index.js [hunk 2]
```diff
@@ -19,12 +19,13 @@ const linkify = (href, options) => createHtmlElement({
...
 
 const getAsDocumentFragment = (string, options) => {
 	const fragment = document.createDocumentFragment();
-	for (const [index, text] of Object.entries(string.split(urlRegex()))) {
-		if (index % 2) { // URLs are always in odd positions
+	const entries = Object.entries(string.split(urlRegex())).map(([k, v]) => [Number(k), v]);
+	for (const [index, text] of entries) {
+		if (index % 2 && (index + 2 < entries.length || entries[entries.length - 1][1] !== '…')) { // URLs are always in odd positions
```
Comment (sindresorhus) [inline index.js]: Move the truncation logic to a `isTruncated` variable for readability.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js [hunk 3]
```diff
@@ -19,12 +19,13 @@ const linkify = (href, options) => createHtmlElement({
...
 const domify = html => document.createRange().createContextualFragment(html);
 
-const getAsString = (string, options) => string.replace(urlRegex(), match => linkify(match, options));
+const getAsString = (string, options) => string.replace(urlRegex(), (match, _, offset) => offset + match.length + 1 === string.length && string.charAt(string.length - 1) === '…' ? match : linkify(match, options));
 
 const getAsDocumentFragment = (string, options) => {
 	const fragment = document.createDocumentFragment();
-	for (const [index, text] of Object.entries(string.split(urlRegex()))) {
-		if (index % 2) { // URLs are always in odd positions
+	const entries = Object.entries(string.split(urlRegex())).map(([k, v]) => [Number(k), v]);
```
Comment (fregante) [inline index.js]:
```suggestion
	const entries = [...string.split(urlRegex()).entries()];
```
I think

File: test.js [hunk 4]
```diff
@@ -147,3 +147,20 @@ test('supports localhost URLs', t => {
...
 });
+
+test('skips truncated URLs', t => {
+	t.is(linkifyUrls('https://github.com/sindresorhus/linkify-…'), 'https://github.com/sindresorhus/linkify-…');
+	t.is(linkifyUrls('https://github.com/sindresorhus/linkify-… and https://github.com/sindresorhus/linkify-…'), '<a href="https://github.com/sindresorhus/linkify-">https://github.com/sindresorhus/linkify-</a>… and https://github.com/sindresorhus/linkify-…');
+	t.is(
+		html(linkifyUrls('See https://github.com/sindresorhus/linkify-urls… and https://github.com/sindresorhus/linkify-…', {
+			type: 'dom',
+		})),
+		html(domify('See <a href="https://github.com/sindresorhus/linkify-urls">https://github.com/sindresorhus/linkify-urls</a>… and https://github.com/sindresorhus/linkify-…')),
```
Comment (fregante) [inline test.js]: It doesn't look like it's working. Neither one should be linkified.

**Review by fregante — COMMENTED** [review]
What do you think about also ignoring URLs that contain …? I'm not sure if the regex already excludes them. Example:

    https://example.com/very-long-…happy-thereafter.html

File: index.js (hunk 1 above)
Comment (fregante) [reply index.js]: I don't think this is the right logic. It should be:

- if next item starts with …, then return match as is

I think the currently logic skips linkifying: `a https://example.com like this…` simply because the whole string ends with … 

I'd include this example in the tests

File: index.js
Comment (fregante) [inline index.js:1]: I'd test for both `"…"` and `"..."`.

An example in the wild: https://news.ycombinator.com/item?id=37245767

(already linkified though)

File: index.js (hunk 1 above)
Comment (Katsute) [reply index.js]: [`d1f54c7` (#40)](https://github.com/sindresorhus/linkify-urls/pull/40/commits/d1f54c794936214a394e5e559c92bf3b29d40e8a)

`offset + match.length + 1 === string.length` checks if the link is directly next to the last character in the input.

Passes test:

https://github.com/Katsute/linkify-urls/blob/d1f54c794936214a394e5e559c92bf3b29d40e8a/test.js#L154

File: index.js (hunk 2 above)
Comment (Katsute) [reply index.js]: [`620ebb2` (#40)](https://github.com/sindresorhus/linkify-urls/pull/40/commits/620ebb28165fb316efaba7518ecae545d291672e)

File: test.js (hunk 4 above)
Comment (Katsute) [reply test.js]: I wasn't sure how to handle that case, websites usually do not truncate in the middle of a text body.

File: index.js (hunk 3 above)
Comment (Katsute) [reply index.js]: [`94a9f85` (#40)](https://github.com/sindresorhus/linkify-urls/pull/40/commits/94a9f8544cd93bbc967de4e1679931f9ee1697ea)

File: index.js
Comment (Katsute) [reply index.js:1]: [`3441ead` (#40)](https://github.com/sindresorhus/linkify-urls/pull/40/commits/3441ead049ac1f24363cc7366803dd4c4a6e561e)

This one is a bit different when checking for `...` since the regex captures this as part of the URL:

* `getAsString` checks if the URL is at the end of the input and ends with `...`

* `getAsDocumentFragment` checks if the URL is the second to last item and ends with `...` and the last item is blank

File: test.js (hunk 4 above)
Comment (fregante) [reply test.js]: Arguments:

- Hacker news is such example, URLs are truncated anywhere they appear (but this happens as a consequence of linkification)
- If `https://example.com/…` is skipped, then `https://example.com/… for example` should probably be skipped too

I'm not confident about this though. From a consistency standpoint they should both be skipped, but in practice I don't know. 

 @sindresorhus?

File: index.js [hunk 5]
```diff
@@ -19,12 +19,15 @@ const linkify = (href, options) => createHtmlElement({
 // Get DOM node from HTML
 const domify = html => document.createRange().createContextualFragment(html);
 
-const getAsString = (string, options) => string.replace(urlRegex(), match => linkify(match, options));
+const isTruncated = (index, entries) => index + 2 === entries.length && (entries[entries.length - 1][1] === '…' || (entries[index][1].endsWith('...') && entries[entries.length - 1][1] === ''));
```
Comment (sindresorhus) [inline index.js]:
```suggestion
const isTruncated = (index, entries) => index + 2 === entries.length && (entries[entries.length - 1][1] === '…' || (entries[index][1].endsWith('…') && entries[entries.length - 1][1] === ''));
```

File: index.js [hunk 6]
```diff
@@ -19,12 +19,15 @@ const linkify = (href, options) => createHtmlElement({
...
+
+const getAsString = (string, options) => string.replace(urlRegex(), (match, _, offset) => (offset + match.length + 1 === string.length && string.endsWith('…')) || (offset + match.length === string.length && string.endsWith('...')) ? match : linkify(match, options));
```
Comment (sindresorhus) [inline index.js]:
```suggestion
const getAsString = (string, options) => string.replace(urlRegex(), (match, _, offset) => (offset + match.length + 1 === string.length && string.endsWith('…')) || (offset + match.length === string.length && string.endsWith('…')) ? match : linkify(match, options));
```

File: test.js (hunk 4 above)
Comment (sindresorhus) [reply test.js]: > If https://example.com/… is skipped, then https://example.com/… for example should probably be skipped too

Agreed

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js (hunk 5 above)
Comment (Katsute) [reply index.js]: Intentional, see https://github.com/sindresorhus/linkify-urls/pull/40#discussion_r1304299859

File: index.js (hunk 6 above)
Comment (Katsute) [reply index.js]: Also intentional, see https://github.com/sindresorhus/linkify-urls/pull/40#discussion_r1304299859

File: index.js (hunk 5 above)
Comment (fregante) [reply index.js]: No this was my suggestion to support both. 

Also this long condition is hard to follow. It should probably be broken up into "if else and return" plus comments.

**Comment by Katsute** [comment]
Skips linkifying any URL followed by `…` or `...`

File: index.js (hunk 5 above)
Comment (Katsute) [reply index.js]: [`f2e5d4c` (#40)](https://github.com/sindresorhus/linkify-urls/pull/40/commits/f2e5d4c0c1b92bc112762eb1f7d5643c04adaeba)

**Review by sindresorhus — APPROVED** [review]

File: index.js [hunk 7]
```diff
@@ -19,12 +19,22 @@ const linkify = (href, options) => createHtmlElement({
 // Get DOM node from HTML
 const domify = html => document.createRange().createContextualFragment(html);
 
-const getAsString = (string, options) => string.replace(urlRegex(), match => linkify(match, options));
+const isTruncated = (url, peek) =>
+	url.endsWith('...') // `...` is a matched by the URL regex
+	|| peek.startsWith('…'); // `…` can follow the match
```
Comment (fregante) [inline index.js:24]: This change slightly extracted the logic into the code itself (the `peek` extraction) but this actually allowed simplifying this feature substantially as well as logic reuse.

File: index.js [hunk 8]
```diff
@@ -19,12 +19,22 @@ const linkify = (href, options) => createHtmlElement({
...
+		? url
+		: linkify(url, options));
 
 const getAsDocumentFragment = (string, options) => {
 	const fragment = document.createDocumentFragment();
-	for (const [index, text] of Object.entries(string.split(urlRegex()))) {
-		if (index % 2) { // URLs are always in odd positions
+	const parts = string.split(urlRegex());
+
+	for (const [index, text] of parts.entries()) {
```
Comment (fregante) [inline index.js:35]: This change dropped one loop `[...entries]` and avoided awkward `[1]` entries access

**Review by fregante — APPROVED** [review]
LGTM now

**Comment by fregante** [comment]
Thanks for the PR @Katsute!
